lxd/storage_volumes.go

```diff
@@ -2729,6 +2729,104 @@ func createStoragePoolVolumeFromBackup(s *state.State, r *http.Request, requestP
 	return operations.OperationResponse(op)
 }
 
+// ctxStorageVolumeDetails is the request.CtxKey corresponding to storageVolumeDetails, which is added to the request
+// context in addStoragePoolVolumeDetailsToRequestContext.
+const ctxStorageVolumeDetails request.CtxKey = "storage-volume-details"
+
+// storageVolumeDetails contains details common to all storage volume requests. A value of this type is added to the
+// request context when addStoragePoolVolumeDetailsToRequestContext is called. We do this to avoid repeated logic when
+// parsing the request details and/or making database calls to get the storage pool or effective project. These fields
+// are required for the storage volume access check, and are subsequently available in the storage volume handlers.
+type storageVolumeDetails struct {
+	volumeName         string
+	volumeTypeName     string
+	volumeType         int
+	pool               storagePools.Pool
+	forwardingNodeInfo *db.NodeInfo
+}
+
+// addStoragePoolVolumeDetailsToRequestContext extracts storageVolumeDetails from the http.Request and adds it to the
+// request context with the ctxStorageVolumeDetails request.CtxKey. Additionally, the effective project of the storage
+// bucket is added to the request context under request.CtxEffectiveProjectName.
+func addStoragePoolVolumeDetailsToRequestContext(s *state.State, r *http.Request) error {
+	var details storageVolumeDetails
+	defer func() {
+		request.SetCtxValue(r, ctxStorageVolumeDetails, details)
```

Comment:
Why do you do this as a defer rather than at the end, as otherwise it maybe set to partial state?

Reply:
We expect a partial state when either the `target` parameter is set, or when the request has already been forwarded (by checking `request.CtxProtocol`). In this case the `forwardingNodeInfo` field is not set. This defer was a convenience really, to make sure it is set in the context. An error is returned if we fail to parse anything.